src/Binding/AbstractBinding.php

```diff
@@ -138,39 +181,43 @@ public function hasParameterValue($parameterName)
     /**
      * {@inheritdoc}
      */
-    public function equals(ResourceBinding $other)
+    public function serialize()
     {
-        if (get_class($other) !== get_class($this)) {
-            return false;
-        }
+        $data = array();
 
-        if ($this->query !== $other->getQuery()) {
-            return false;
-        }
+        $this->preSerialize($data);
 
-        if ($this->type !== $other->getType()) {
-            return false;
-        }
+        return serialize($data);
+    }
 
-        if ($this->language !== $other->getLanguage()) {
-            return false;
-        }
+    /**
+     * {@inheritdoc}
+     */
+    public function unserialize($serialized)
+    {
+        $data = unserialize($serialized);
 
-        // The local parameters are sorted by key. Sort before comparing to
-        // prevent false negatives.
-        $otherParameterValues = $other->getParameterValues();
-        ksort($otherParameterValues);
+        $this->postUnserialize($data);
+    }
 
-        if ($this->parameterValues !== $otherParameterValues) {
-            return false;
-        }
+    protected function preSerialize(array &$data)
+    {
+        $data[] = $this->typeName;
+        $data[] = $this->userParameterValues;
+        $data[] = $this->uuid->toString();
+    }
 
-        return true;
+    protected function postUnserialize(array $data)
+    {
+        $this->uuid = Uuid::fromString(array_pop($data));
```

Comment:
passing the array by value in the ``postUnserialize`` function and then changing it by reference here causes a reference mismatch forcing PHP to duplicate the array in memory (the caller still holds a reference to the array in its ``$data`` when the method is called.

There is 2 solutions:

- inline the ``unserialize`` call, removing the ``$data``, which may solve the reference mismatch (not sure how PHP handles releasing the reference to the zval, but it should be OK as there is nothing else than the argument referencing it)
- read values from the array instead of altering the array here (this will avoid the reference mismatch for sure).

Given that you only have 3 elements in the array, it may not be too consuming to duplicate the array here. But this code could be taken as example for other places dealing with bigger arrays, in which case it would become an issue.